File: FitBuddy/MainViewControllers/LogbookEntryViewController.m
//
//  LogbookEntryViewController.m
//  GymBuddy
//
//  Created by John Neyer on 2/19/12.
//  Copyright (c) 2012 jneyer.com. All rights reserved.
//

#import "LogbookEntryViewController.h"
#import "CoreDataHelper.h"
#import "FitBuddyMacros.h"

@implementation LogbookEntryViewController

@synthesize logbookEntry = _logbookEntry;
@synthesize entryDate = _entryDate;
@synthesize entryName = _entryName;
@synthesize slotOneLabel = _slotOneLabel;
@synthesize slotTwoLabel = _slotTwoLabel;
@synthesize slotThreeLabel = _slotThreeLabel;
@synthesize colOneDate = _colOneDate;
@synthesize colTwoDate = _colTwoDate;
@synthesize slotOneColOne = _slotOneColOne;
@synthesize slotTwoColOne = _slotTwoColOne;
@synthesize slotThreeColOne = _slotThreeColOne;
@synthesize slotOneColTwo = _slotOneColTwo;
@synthesize slotTwoColTwo = _slotTwoColTwo;
@synthesize slotThreeColTwo = _slotThreeColTwo;

- (void)viewDidLoad
{
    [super viewDidLoad];
    self.navigationItem.titleView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:kFITBUDDY]];
    
}

-(NSString *) tinyDateFormat: (NSDate *) date
                      format:(NSString *) format
{
    NSDateFormatter *formatter = [[NSDateFormatter alloc] init];
    [formatter setDateFormat:format];
    return [formatter stringFromDate:date];
}

-(void) loadFormDataFromLogbook
{
    // Set the data
    self.entryDate.text = [self tinyDateFormat:self.logbookEntry.date format:@"dd MMM yyyy"];
    
    self.entryName.text = self.logbookEntry.exercise_name;
    
    if (self.logbookEntry.pace || self.logbookEntry.distance || self.logbookEntry.duration)
    {
        self.slotOneLabel.text = @"Pace";
        self.slotOneColOne.text = self.logbookEntry.pace;
        
        self.slotTwoLabel.text = @"Duration";
        self.slotTwoColOne.text = self.logbookEntry.duration;
        
        self.slotThreeLabel.text = @"Distance";
        self.slotThreeColOne.text = self.logbookEntry.distance;
    }
    else
    {
        self.slotOneLabel.text = @"Weight";
        self.slotOneColOne.text = self.logbookEntry.weight;
        
        self.slotTwoLabel.text = @"Sets";
        self.slotTwoColOne.text = self.logbookEntry.sets;
        
        self.slotThreeLabel.text = @"Reps";
        self.slotThreeColOne.text = self.logbookEntry.reps;   
    }
    
    self.colOneDate.text = [self tinyDateFormat:self.logbookEntry.date format:@"MM/dd/yy"];
}


-(void) loadLastEntryFromLogbook
{
    NSString *exerciseName = self.logbookEntry.exercise_name;
    NSDate *entryDate = [self.logbookEntry.date copy];
    NSFetchedResultsController *frc;
    
    NSFetchRequest *request = [NSFetchRequest fetchRequestWithEntityName:LOGBOOK_TABLE];
    request.sortDescriptors = [NSArray arrayWithObject:[NSSortDescriptor sortDescriptorWithKey:@"date" ascending:YES]];
    request.predicate = [NSPredicate predicateWithFormat:@"(exercise_name = %@) AND (date < %@) AND (completed = %@)", 
                         exerciseName, entryDate, [NSNumber numberWithBool:YES]];
    
    frc = [[NSFetchedResultsController alloc] initWithFetchRequest:request 
                                              managedObjectContext:[[AppDelegate sharedAppDelegate] managedObjectContext]
                                                sectionNameKeyPath:nil 
                                                         cacheName:nil];
    NSError *error;
    [frc performFetch:&error];
    NSLog(@"Retrieving logbook");
    
    NSArray *array = [frc fetchedObjects];
    LogbookEntry *theEntry;
    
    if (array && array.count > 0)
    {
        theEntry = (LogbookEntry *)[array lastObject];
        
        self.colTwoDate.text = [self tinyDateFormat:theEntry.date format:@"MM/dd/yy"];
        
        if (theEntry.pace || theEntry.distance || theEntry.duration)
        {
            self.slotOneColTwo.text = theEntry.pace;
            self.slotTwoColTwo.text = theEntry.duration;
            self.slotThreeColTwo.text = theEntry.distance;
        }
        else
        {
            self.slotOneColTwo.text = theEntry.weight;
            self.slotTwoColTwo.text = theEntry.sets;
            self.slotThreeColTwo.text = theEntry.reps;   
        }  
        
    }
    else
    {
        self.colTwoDate.text = [self tinyDateFormat:theEntry.date format:@"--/--/--"];
        self.slotOneColTwo.text = @"-";
        self.slotTwoColTwo.text = @"-";
        self.slotThreeColTwo.text = @"-"; 
    }
}

-(void) viewWillAppear:(BOOL)animated
{
    // Data stuff
    [self loadFormDataFromLogbook];
    
    // Last Entry
    [self loadLastEntryFromLogbook];
}

- (void)viewDidUnload {
    [self setEntryDate:nil];
    [self setEntryName:nil];
    [self setSlotOneLabel:nil];
    [self setSlotTwoLabel:nil];
    [self setSlotThreeLabel:nil];
    [self setColOneDate:nil];
    [self setColTwoDate:nil];
    [self setSlotOneColOne:nil];
    [self setSlotTwoColOne:nil];
    [self setSlotThreeColOne:nil];
    [self setSlotOneColTwo:nil];
    [self setSlotTwoColTwo:nil];
    [self setSlotThreeColTwo:nil];
    [super viewDidUnload];
}
@end

File: controllers/LogbookEntryViewController.swift
//
//  LogbookEntryViewController.swift
//  FitBuddy
//
//  Created by john.neyer on 5/16/15.
//  Copyright (c) 2015 jneyer.com. All rights reserved.
//

import Foundation
import UIKit
import FitBuddyModel
import FitBuddyCommon

class LogbookEntryViewController: UIViewController {
    
    var logbookEntry: LogbookEntry? = nil
    
    @IBOutlet weak var entryDate: UILabel!
    @IBOutlet weak var entryName: UILabel!
    @IBOutlet weak var slotOneLabel: UILabel!
    @IBOutlet weak var slotTwoLabel: UILabel!
    @IBOutlet weak var slotThreeLabel: UILabel!
    @IBOutlet weak var colOneDate: UILabel!
    @IBOutlet weak var colTwoDate: UILabel!
    @IBOutlet weak var slotOneColOne: UILabel!
    @IBOutlet weak var slotTwoColOne: UILabel!
    @IBOutlet weak var slotThreeColOne: UILabel!
    @IBOutlet weak var slotOneColTwo: UILabel!
    @IBOutlet weak var slotTwoColTwo: UILabel!
    @IBOutlet weak var slotThreeColTwo: UILabel!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.navigationItem.titleView = UIImageView(image: UIImage(named: FBConstants.kFITBUDDY))
    }
    
    func loadFormDataFromLogbook () {
        
        // Set the data
        self.entryDate.text = FitBuddyUtils.dateFromNSDate(self.logbookEntry!.date, format: "dd MMM yyyy")
        self.entryName.text = self.logbookEntry!.exercise_name
        
        if !self.logbookEntry!.pace!.isEmpty {
            self.slotOneLabel.text = "Pace"
            self.slotOneColOne.text = self.logbookEntry!.pace
            
            self.slotTwoLabel.text = "Duration"
            self.slotTwoColOne.text = self.logbookEntry!.duration
            
            self.slotThreeLabel.text = "Distance"
            self.slotThreeColOne.text = self.logbookEntry!.distance
        }
        else
        {
            self.slotOneLabel.text = "Weight"
            self.slotOneColOne.text = self.logbookEntry!.weight
        
            self.slotTwoLabel.text = "Sets"
            self.slotTwoColOne.text = self.logbookEntry!.sets
            
            self.slotThreeLabel.text = "Reps"
            self.slotThreeColOne.text = self.logbookEntry!.reps
        }
        
        self.colOneDate.text = FitBuddyUtils.shortDateFromNSDate(self.logbookEntry!.date)
        
    }
    
    
    func loadLastEntryFromLogbook () {
        
        let exerciseName = self.logbookEntry!.exercise_name
        let entryDate = self.logbookEntry!.date.copy() as! NSDate
        
        let request = NSFetchRequest(entityName: FBConstants.LOGBOOK_TABLE)
        request.sortDescriptors = [NSSortDescriptor(key: "date", ascending: true)]
        request.predicate = NSPredicate(format: "(exercise_name = %@) AND (date < %@) AND (completed = %@)", argumentArray: [exerciseName, entryDate, 1])
        
        let frc: NSFetchedResultsController = NSFetchedResultsController(fetchRequest: request, managedObjectContext: AppDelegate.sharedAppDelegate().managedObjectContext!, sectionNameKeyPath: nil, cacheName: nil)
        
        var error: NSError? = nil
        frc.performFetch(&error)
        
        NSLog("Retrieving logbook")
        let array = frc.fetchedObjects
        
        var theEntry: LogbookEntry?
        
        if array != nil && array?.count > 0 {
            
            theEntry = (array!.last as! LogbookEntry)
            
            self.colTwoDate.text = FitBuddyUtils.shortDateFromNSDate(theEntry?.date)
            
            if !theEntry!.pace!.isEmpty {
                self.slotOneColTwo.text = theEntry!.pace
                self.slotTwoColTwo.text = theEntry!.duration
                self.slotThreeColTwo.text = theEntry!.distance
            }
            else
            {
                self.slotOneColTwo.text = theEntry!.weight
                self.slotTwoColTwo.text = theEntry!.sets
                self.slotThreeColTwo.text = theEntry!.reps
            }
            
        }
        else
        {
            self.colTwoDate.text = "--/--/--"
            self.slotOneColTwo.text = "-"
            self.slotTwoColTwo.text = "-"
            self.slotThreeColTwo.text = "-"
        }
    }
    
    override func viewWillAppear(animated: Bool) {
        self.loadFormDataFromLogbook()
        self.loadLastEntryFromLogbook()
    }
    
}